<a href="https://colab.research.google.com/github/ashpou1/data-processing-4313/blob/main/SQL%20Essentials.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ASHOK SHARMA POUDEL  
# 218380  
# DATA PROCESSING VISUALIZATION  4313 
# DR. NICHOLAS JACOB

##Learning Some SQL with BigQuery
The first bit below comes directly from Google, you'll need to do each of those to be successful in getting this document to work.

##Cloud Resource Manager to Create a Cloud Platform project
##Enable BigQuery APIs for the project.

In [1]:
from google.colab import auth
auth.authenticate_user()
print('Authenticated')

Authenticated


I will use austin_bikeshare data set of google trends.

## TOTAL ROWS
I will use my project id and count my total rows in data frame of bikeshare trips

In [2]:
%%bigquery --project data-viz-338818
SELECT 
  COUNT(*) as total_rows
FROM `bigquery-public-data.austin_bikeshare.bikeshare_trips`

total_rows
0     1424786

$1424786$ is a big data. 

##LONGEST TRIP IN MINUTES
Lets find out longest trip in data set

In [3]:
%%bigquery --project data-viz-338818
SELECT 
  MAX(duration_minutes) as Longest_trip_in_minutes
FROM `bigquery-public-data.austin_bikeshare.bikeshare_trips`

Longest_trip_in_minutes
0                    34238

longest trip in data is $34238$ minutes

##AVERAGE TIME FOR TRIP
lets compute average time for trip in data set. we will use start station name to end station name and duration as minutes. 

In [4]:
%%bigquery --project data-viz-338818
SELECT 
  start_station_name, 
  end_station_name,
  AVG(duration_minutes) as Average_time_for_trip
FROM `bigquery-public-data.austin_bikeshare.bikeshare_trips`
WHERE start_station_name is not null
GROUP BY start_station_name, end_station_name

start_station_name  ... Average_time_for_trip
0                                       Zilker Park West  ...             22.636364
1                                Toomey Rd @ South Lamar  ...             48.985782
2                                       Waller & 6th St.  ...             38.334477
3                                             Pease Park  ...             30.497006
4                                       Waller & 6th St.  ...             18.060000
...                                                  ...  ...                   ...
13962  South Congress & Barton Springs at the Austin ...  ...             29.971429
13963  South Congress & Barton Springs at the Austin ...  ...             49.000000
13964  South Congress & Barton Springs at the Austin ...  ...             39.500000
13965  South Congress & Barton Springs at the Austin ...  ...             42.642857
13966  South Congress & Barton Springs at the Austin ...  ...             47.509804

[13967 rows x 3 columns]

We got as many as 13967 average trips time which starts from  $22.636364$ to $47.509804$. these trips represent aveage time of trip from same start and end points. 

# SQL Essentials

lets have a general look data table limiting data $10$

In [5]:
%%bigquery --project data-viz-338818
SELECT *
FROM `bigquery-public-data.austin_bikeshare.bikeshare_trips`
LIMIT 10


trip_id  ... duration_minutes
0  9900289692  ...               39
1  9900285987  ...               31
2  9900285989  ...               31
3  9900285991  ...               30
4  9900286140  ...               19
5  9900286143  ...               17
6  9900286171  ...                6
7  9900286214  ...                0
8  9900286540  ...               12
9    13575843  ...               47

[10 rows x 9 columns]

# Compute the average time for a trip based on starting point


lets use start station name and time of trip to get average time.

In [9]:
%%bigquery --project data-viz-338818
SELECT 
  start_station_name, 
  AVG(duration_minutes) as Average_time_for_trip
FROM `bigquery-public-data.austin_bikeshare.bikeshare_trips`
WHERE start_station_name is not null
GROUP BY start_station_name

start_station_name  Average_time_for_trip
0                         Zilker Park West              27.168259
1                  Toomey Rd @ South Lamar              28.115798
2          State Capitol @ 14th & Colorado              30.421842
3                         Waller & 6th St.              23.265843
4                               Pease Park              28.648889
..                                     ...                    ...
188  Republic Square @ Guadalupe & 4th St.              23.067313
189                             3rd & West              19.937466
190                               3rd/West              34.152547
191                          Nueces & 26th              16.805353
192                            26th/Nueces              24.521670

[193 rows x 2 columns]

Among $193$ starting points we have averege time of a trip from $24.52167$ to $27.168259$  And we can also know among $1424786$ (which is our total row or we can say total trips) we only have $193$ starting points. 

# Compute how many trips start at each starting point.

From our general view of data table we will select start station name and use bigquery to count total trips from each starting point.  

In [7]:
%%bigquery --project data-viz-338818
SELECT 
  start_station_name,
  COUNT(*) as total_trips
FROM `bigquery-public-data.austin_bikeshare.bikeshare_trips`
GROUP BY start_station_name
ORDER BY total_trips DESC

start_station_name  total_trips
0        21st & Speedway @PCL        72799
1        Riverside @ S. Lamar        40635
2    City Hall / Lavaca & 2nd        36520
3              2nd & Congress        35307
4        Rainey St @ Cummings        34758
..                        ...          ...
188           Marketing Event            4
189             Eeyore's 2018            2
190                    Stolen            1
191             Eeyore's 2017            1
192     cesar Chavez/Congress            1

[193 rows x 2 columns]

So, we got $193$ starting point and from $1$ to $72799$ from a individual single starting point.